components/site/src/link_checking.rs

```diff
@@ -63,7 +63,10 @@ pub fn check_internal_links_with_anchors(site: &Site) -> Result<()> {
             let page = library
                 .get_page(&full_path)
                 .expect("Couldn't find section in check_internal_links_with_anchors");
-            !page.has_anchor(anchor)
+
+            !(page.has_anchor(anchor)||
+              site.config.link_checker.heuristic_link_check &&
+              page.has_anchor_id(anchor))
```

Comment:
Is regex created for each page? Would caching it make processing faster?

Reply:
For each anchor and each page. I would guess that it would only make a minor difference in most cases.

Reply:
I think caching the regex wouldn't be as useful because the regex pattern contains the text of each anchor, so the likelihood any particular regex would be needed again is slim.

Unless the method was flipped around, so there was a single cached regex behind a mutex, with a pattern like `r"(?i) (id|name) *= *"[^"]"? "` (case insesitively match _all_ name or id attributes (I used only double quoting for simplicity)).  Then there's only one regex, which can find all matches in a given page, then you can check those matches for the anchor in question.

I think there's some chance that would be faster than creating so many fresh regexes.

Reply:
@mwcz I would agree that this might be faster. Then I thought regexs would be faster than exact string matching and was wrong. We don't have a good test realistic test case for any kind of real world benchmarking. My natural inclination would be to leave the simplest implementation till it's clear that this is a problem.

Reply:
For sure, +1 to keeping it simple at first.

For benchmarking, I might be able to help with that.  The site I'm starting to build with zola has 130,000 pages of markdown, with a lot of name/id anchors, so it would be highly sensitive to any performance issues.  I can run benchmarks myself with that content.  Maybe if I can find a way to scrub the content (lorem-ipsumify it but keep the markdown layout) then I could share it.

Reply:
@mwcz That would be a good benchmark but only to see it it this code does not affect it at all. This code only runs if the current link checker fails. So if your current site builds (i.e. all the links work), then this code should never run. That's a useful thing to check, but doesn't give any indication of the performance of this code. I'm assuming your anchors here are either zola generated section slugs or manually specified anchors.

What you need could do is take half the markdown files, turn them into header/footer-less HTML with pandoc or equivalent. Now you should have 65,000 markdown files with pointers to 65,000 HTML files, so half of your outlinks would be to XML specified anchors and therefore use this code.

Quite how we would interpret the results I don't know. This check is surely slower than checking for markdown anchors which have to be parsed out anyway. So, it's not clear what would be an acceptable slowdown.

Reply:
The fastest implementation would be integrated into the markdown=>HTML convertor, where it's easy to collect all ids at the conversion stage and then to finally check at the end.

Reply:
@biodranik Who knows what would be fastest, without a good test. Collecting all the ids/links would potentially mean checking a lot more links, so it might be slower. Of course, it does mean that the link checker could check all links. At the moment HTML -> HTML links that are broken would not be picked up.

It's a question of what the link checker is there for. I think it's fine that it is limited to markdown links, so long as it does not preclude me putting in HTML anchors.